src/write_sensu.c

```diff
@@ -0,0 +1,1230 @@
+
+static void free_str_list(struct str_list *strs) /* {{{ */
+{
+	int i;
+	for (i=0; i<strs->nb_strs; i++)
+		free(strs->strs[i]);
+	free(strs->strs);
+}
+/* }}} void free_str_list */
+
+static int sensu_connect(struct sensu_host *host) /* {{{ */
+{
+	int			 e;
+	struct addrinfo		*ai, hints;
+	char const		*node;
+	char const		*service;
+
+	// Resolve the target if we haven't done already
+	if (!(host->flags & F_READY)) {
+		memset(&hints, 0, sizeof(hints));
+		memset(&service, 0, sizeof(service));
+		host->res = NULL;
+		hints.ai_family = AF_INET;
+		hints.ai_socktype = SOCK_STREAM;
+#ifdef AI_ADDRCONFIG
+		hints.ai_flags |= AI_ADDRCONFIG;
+#endif
+
+		node = (host->node != NULL) ? host->node : SENSU_HOST;
+		service = (host->service != NULL) ? host->service : SENSU_PORT;
+
+		if ((e = getaddrinfo(node, service, &hints, &(host->res))) != 0) {
+			ERROR("write_sensu plugin: Unable to resolve host \"%s\": %s",
+					node, gai_strerror(e));
+			return -1;
+		}
+		DEBUG("write_sensu plugin: successfully resolved host/port: %s/%s",
+				node, service);
+		host->flags |= F_READY;
+	}
+
+	struct linger so_linger;
+	host->s = -1;
+	for (ai = host->res; ai != NULL; ai = ai->ai_next) {
+		// create the socket
+		if ((host->s = socket(ai->ai_family,
+				      ai->ai_socktype,
+				      ai->ai_protocol)) == -1) {
+			continue;
+		}
+
+		// Set very low close() lingering
+		so_linger.l_onoff = 1;
+		so_linger.l_linger = 3;
+		if (setsockopt(host->s, SOL_SOCKET, SO_LINGER, &so_linger, sizeof so_linger) != 0)
+			WARNING("write_sensu plugin: failed to set socket close() lingering");
+
+		// connect the socket
+		if (connect(host->s, ai->ai_addr, ai->ai_addrlen) != 0) {
+			close(host->s);
+			host->s = -1;
+			continue;
+		}
+		DEBUG("write_sensu plugin: connected");
+		break;
+	}
+
+	if (host->s < 0) {
+		WARNING("write_sensu plugin: Unable to connect to sensu client");
+		return -1;
+	}
+	return 0;
+} /* }}} int sensu_connect */
+
+static void sensu_close_socket(struct sensu_host *host) /* {{{ */
+{
+	if (host->s != -1)
+		close(host->s);
+	host->s = -1;
+
+} /* }}} void sensu_close_socket */
+
+static char *build_json_str_list(const char *tag, struct str_list const *list) /* {{{ */
+{
+	int res;
+	char *ret_str;
+	char *temp_str;
+	int i;
+	if (list->nb_strs == 0) {
+		ret_str = malloc(sizeof(char));
+		if (ret_str == NULL) {
+			ERROR(alloc_err);
+			return NULL;
+		}
+		ret_str[0] = '\0';
+	}
+
+	res = asprintf(&temp_str, "\"%s\": [\"%s\"", tag, list->strs[0]);
+	if (res == -1) {
+		ERROR(alloc_err);
+		return NULL;
+	}
+	for (i=1; i<list->nb_strs; i++) {
+		res = asprintf(&ret_str, "%s, \"%s\"", temp_str, list->strs[i]);
+		free(temp_str);
+		if (res == -1) {
+			ERROR(alloc_err);
+			return NULL;
+		}
+		temp_str = ret_str;
+	}
+	res = asprintf(&ret_str, "%s]", temp_str);
+	free(temp_str);
+	if (res == -1) {
+		ERROR(alloc_err);
+		return NULL;
+	}
+
+	return ret_str;
+} /* }}} char *build_json_str_list*/
+
+int format_name2(char *ret, int ret_len,
```

Comment:
Would it make sense renaming this function to something prefixed by `sensu_` ?

Also this function is the only part of the file indented by whitspaces instead of tabs.